ipnsfs/system.go

```diff
@@ -0,0 +1,285 @@
+	eventlog "github.com/jbenet/go-ipfs/thirdparty/eventlog"
+)
+
+var log = eventlog.Logger("ipnsfs")
+
+var ErrIsDirectory = errors.New("error: is a directory")
+
+var ErrNoSuch = errors.New("no such file or directory")
+
+// Filesystem is the writeable fuse filesystem structure
+type Filesystem struct {
+	dserv dag.DAGService
+
+	nsys namesys.NameSystem
+
+	pins pin.Pinner
+
+	roots map[string]*KeyRoot
+}
+
+func NewFilesystem(ctx context.Context, ds dag.DAGService, nsys namesys.NameSystem, pins pin.Pinner, keys ...ci.PrivKey) (*Filesystem, error) {
+	roots := make(map[string]*KeyRoot)
+	fs := &Filesystem{
+		roots: roots,
+		nsys:  nsys,
+		dserv: ds,
+		pins:  pins,
+	}
+	for _, k := range keys {
+		pkh, err := k.GetPublic().Hash()
+		if err != nil {
+			return nil, err
+		}
+
+		root, err := fs.NewKeyRoot(ctx, k)
+		if err != nil {
+			return nil, err
+		}
+		roots[u.Key(pkh).Pretty()] = root
+	}
+
+	return fs, nil
+}
+
+func (fs *Filesystem) Open(tpath string, mode int) (File, error) {
+	pathelem := strings.Split(tpath, "/")
+	r, ok := fs.roots[pathelem[0]]
+	if !ok {
+		return nil, ErrNoSuch
+	}
+
+	return r.Open(pathelem[1:], mode)
+}
+
+func (fs *Filesystem) Close() error {
+	for _, r := range fs.roots {
+		err := r.Publish(context.TODO())
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (fs *Filesystem) GetRoot(name string) (*KeyRoot, error) {
+	r, ok := fs.roots[name]
+	if ok {
+		return r, nil
+	}
+	return nil, ErrNoSuch
+}
+
+type NodeType int
+
+const (
+	TFile NodeType = iota
+	TDir
+)
+
+type FSNode interface {
+	GetNode() (*dag.Node, error)
+	Type() NodeType
+}
+
+// KeyRoot represents the root of a filesystem tree pointed to by a given keypair
+type KeyRoot struct {
+	key ci.PrivKey
+
+	// node is the merkledag node pointed to by this keypair
+	node *dag.Node
+
+	// A pointer to the filesystem to access components
+	fs *Filesystem
+
+	// val represents the node pointed to by this key. It can either be a File or a Directory
+	val FSNode
+
+	repub *Republisher
+}
+
+func (fs *Filesystem) NewKeyRoot(parent context.Context, k ci.PrivKey) (*KeyRoot, error) {
+	hash, err := k.GetPublic().Hash()
+	if err != nil {
+		return nil, err
+	}
+
+	name := u.Key(hash).Pretty()
+
+	root := new(KeyRoot)
+	root.key = k
+	root.fs = fs
+
+	ctx, cancel := context.WithCancel(parent)
+	defer cancel()
+
+	pointsTo, err := fs.nsys.Resolve(ctx, name)
+	if err != nil {
+		err = namesys.InitializeKeyspace(ctx, fs.dserv, fs.nsys, fs.pins, k)
+		if err != nil {
+			return nil, err
+		}
+
+		pointsTo, err = fs.nsys.Resolve(ctx, name)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	mnode, err := fs.dserv.Get(pointsTo)
+	if err != nil {
+		return nil, err
+	}
+
+	root.node = mnode
+
+	root.repub = NewRepublisher(root, time.Millisecond*300, time.Second*3)
+	go root.repub.Run(parent)
+
+	pbn, err := ft.FromBytes(mnode.Data)
+	if err != nil {
+		log.Error("IPNS pointer was not unixfs node")
```

Comment:
this has to be for now since we dont have a defined way of reading other dag formats.